Able to create multiple URL's with devfile with the same name

/kind bug

<!--

Welcome! - We kindly ask you to:

  1. Fill out the issue template below 
  2. Use the Google group if you have a question rather than a bug or feature request.

The group is at: https://groups.google.com/forum/#!forum/odo-users

Thanks for understanding, and for contributing to the project!

-->

## What versions of software are you using?

**Operating System:** Linux

**Output of `odo version`:** Master

## How did you run odo exactly?

```
~/openshift/springboot  master ✗                                                                                                                                            98d ⚑ ◒  
▶ odo url create --host e8ca.engint.REDACTED.com
 ✓  URL myspring-8080.e8ca.engint.REDACTED.com created for component: myspring

To apply the URL configuration changes, please use `odo push`

~/openshift/springboot  master ✗                                                                                                                                            98d ⚑ ◒  
▶ odo url create --host e8ca.engint.REDACTED.com
 ✓  URL myspring-8080.e8ca.engint.REDACTED.com created for component: myspring

To apply the URL configuration changes, please use `odo push`

~/openshift/springboot  master ✗                                                                                                                                            98d ⚑ ◒  
▶ odo url list
Found the following URLs for component myspring
NAME              URL                                               PORT     SECURE
myspring-8080     http://myspring-8080.api.engint.REDACTED.com     8080     false
myspring-8080     http://myspring-8080.api.engint.REDACTED.com     8080     false
myspring-8080     http://myspring-8080.api.engint.REDACTED.com     8080     false

~/openshift/springboot  master ✗                                                                                                                                            98d ⚑ ◒  
▶ odo url delete myspring-8080
? Are you sure you want to delete the url myspring-8080 Yes
panic: runtime error: slice bounds out of range [3:1]

goroutine 1 [running]:
github.com/openshift/odo/pkg/envinfo.(*EnvSpecificInfo).DeleteURL(0xc000b7fe50, 0x7ffd4c13d092, 0xd, 0x0, 0x0)                                                                       
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/pkg/envinfo/envinfo.go:225 +0x4fc                                                                            
github.com/openshift/odo/pkg/odo/cli/url.(*URLDeleteOptions).Run(0xc000739b00, 0x0, 0x0)
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/pkg/odo/cli/url/delete.go:113 +0x73                                                                          
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun(0x20aa900, 0xc000739b00, 0xc0006ce480, 0xc0007e3880, 0x1, 0x1)                                                         
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:31 +0x13c                                                              
github.com/openshift/odo/pkg/odo/cli/url.NewCmdURLDelete.func1(0xc0006ce480, 0xc0007e3880, 0x1, 0x1)                                                                                 
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/pkg/odo/cli/url/delete.go:149 +0x5e                                                                          
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).execute(0xc0006ce480, 0xc0007e3860, 0x1, 0x1, 0xc0006ce480, 0xc0007e3860)                                          
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:702 +0x285                                                          
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc00076ed80, 0x1e94f70, 0xc00004af60, 0x0)                                                               
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:783 +0x2c9                                                          
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).Execute(...)
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:736                                                                 
main.main()
        /home/wikus/syncthing/files/dev/go/src/github.com/openshift/odo/cmd/odo/odo.go:75 +0x460                                                                                     


```


## Actual behavior

Error out when deleting the URL as well as being able to push a URL even though there is already one with the same name.

## Expected behavior


Being able to delete the URL as well as an error happening when trying to create a URL that has the same name.

## Any logs, error output, etc?

--- a/pkg/kclient/fake/ingress.go
+++ b/pkg/kclient/fake/ingress.go
@@ -38,7 +38,7 @@ func GetIngressListWithMultiple(componentName string) *extensionsv1.IngressList 
 						labels.URLLabel:                "example-1",
 					},
 				},
-				Spec: *kclient.GenerateIngressSpec(kclient.IngressParameter{ServiceName: "example-1", PortNumber: intstr.FromInt(8080)}),
+				Spec: *kclient.GenerateIngressSpec(kclient.IngressParameter{ServiceName: "example-1", PortNumber: intstr.FromInt(9090)}),
 			},
 		},
 	}

--- a/pkg/url/url.go
+++ b/pkg/url/url.go
@@ -575,6 +575,7 @@ func Push(client *occlient.Client, kClient *kclient.Client, parameters PushParam
 		for _, url := range urlList.Items {
 			urlCLUSTER[url.Name] = URL{
 				Spec: URLSpec{
+					Host:    url.Spec.Rules[0].Host,
 					Port:    int(url.Spec.Rules[0].HTTP.Paths[0].Backend.ServicePort.IntVal),
 					urlKind: envinfo.INGRESS,
 				},
@@ -603,7 +604,21 @@ func Push(client *occlient.Client, kClient *kclient.Client, parameters PushParam
 	// find URLs to delete
 	for urlName, urlSpec := range urlCLUSTER {
 		val, ok := urlLOCAL[urlName]
-		if !ok {
+
+		configMismatch := false
+		if ok {
+			// since the host stored in an ingress
+			// is the combination of name and host of the url
+			if val.Spec.urlKind == envinfo.INGRESS {
+				val.Spec.Host = fmt.Sprintf("%v.%v", urlName, val.Spec.Host)
+			}
+			if !reflect.DeepEqual(val.Spec, urlSpec.Spec) {
+				configMismatch = true
+				glog.V(4).Infof("config and cluster mismatch for url %s", urlName)
+			}
+		}
+
+		if !ok || configMismatch {
 			if urlSpec.Spec.urlKind == envinfo.INGRESS && kClient == nil {
 				continue
 			}
@@ -614,11 +629,8 @@ func Push(client *occlient.Client, kClient *kclient.Client, parameters PushParam
 			}
 			log.Successf("URL %s successfully deleted", urlName)
 			urlChange = true
+			delete(urlCLUSTER, urlName)
 			continue
-		} else {
-			if !reflect.DeepEqual(val.Spec, urlSpec.Spec) {
-				return errors.Errorf("config mismatch for URL with the same name %s", val.Name)
-			}
 		}
 	}
 

--- a/pkg/url/url.go
+++ b/pkg/url/url.go
@@ -614,7 +614,7 @@ func Push(client *occlient.Client, kClient *kclient.Client, parameters PushParam
 			}
 			if !reflect.DeepEqual(val.Spec, urlSpec.Spec) {
 				configMismatch = true
-				glog.V(4).Infof("config and cluster mismatch for url %s", urlName)
+				klog.V(4).Infof("config and cluster mismatch for url %s", urlName)
 			}
 		}
 

--- a/pkg/url/url_test.go
+++ b/pkg/url/url_test.go
@@ -1081,9 +1081,26 @@ func TestPush(t *testing.T) {
 					*fake.GetSingleIngress("example-local-0", "nodejs"),
 				},
 			},
-			createdURLs: []URL{},
-			deletedURLs: []URL{},
-			wantErr:     true,
+			createdURLs: []URL{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "example-local-0",
+					},
+					Spec: URLSpec{
+						Port:    8080,
+						Secure:  false,
+						urlKind: envinfo.ROUTE,
+					},
+				},
+			},
+			deletedURLs: []URL{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "example-local-0",
+					},
+				},
+			},
+			wantErr: false,
 		},
 		{
 			name:            "url with same name exists on config and cluster but with different specs",
@@ -1103,9 +1120,26 @@ func TestPush(t *testing.T) {
 				},
 			},
 			returnedIngress: &extensionsv1.IngressList{},
-			createdURLs:     []URL{},
-			deletedURLs:     []URL{},
-			wantErr:         true,
+			createdURLs: []URL{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "example-local-0-app",
+					},
+					Spec: URLSpec{
+						Port:    8080,
+						Secure:  false,
+						urlKind: envinfo.ROUTE,
+					},
+				},
+			},
+			deletedURLs: []URL{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "example-local-0-app",
+					},
+				},
+			},
+			wantErr: false,
 		},
 
 		{

--- a/pkg/url/url_test.go
+++ b/pkg/url/url_test.go
@@ -777,11 +777,33 @@ func TestPush(t *testing.T) {
 				},
 				{
 					Name:   "example-1",
-					Port:   8080,
+					Port:   9090,
 					Secure: false,
 				},
 			},
 			returnedRoutes: &routev1.RouteList{},
+			createdURLs: []URL{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "example-app",
+					},
+					Spec: URLSpec{
+						Port:    8080,
+						Secure:  false,
+						urlKind: envinfo.ROUTE,
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "example-1-app",
+					},
+					Spec: URLSpec{
+						Port:    9090,
+						Secure:  false,
+						urlKind: envinfo.ROUTE,
+					},
+				},
+			},
 		},
 		{
 			name:            "0 url on local config and 2 on openshift cluster",
@@ -838,6 +860,27 @@ func TestPush(t *testing.T) {
 					},
 				},
 			},
+		},
+		{
+			name:            "2 url on local config and openshift cluster are in sync",
+			componentName:   "nodejs",
+			applicationName: "app",
+			args:            args{isRouteSupported: true},
+			existingConfigURLs: []config.ConfigURL{
+				{
+					Name:   "example",
+					Port:   8080,
+					Secure: false,
+				},
+				{
+					Name:   "example-1",
+					Port:   9100,
+					Secure: false,
+				},
+			},
+			returnedRoutes: testingutil.GetRouteListWithMultiple("nodejs", "app"),
+			deletedURLs:    []URL{},
+			createdURLs:    []URL{},
 		},
 
 		{
@@ -973,6 +1016,31 @@ func TestPush(t *testing.T) {
 					},
 				},
 			},
+		},
+		{
+			name:          "2 urls on env file and openshift cluster are in sync",
+			componentName: "wildfly",
+			args:          args{isRouteSupported: true, isExperimentalModeEnabled: true},
+			existingEnvInfoURLs: []envinfo.EnvInfoURL{
+				{
+					Name:   "example-0",
+					Port:   8080,
+					Secure: false,
+					Host:   "com",
+					Kind:   envinfo.INGRESS,
+				},
+				{
+					Name:   "example-1",
+					Port:   9090,
+					Secure: false,
+					Host:   "com",
+					Kind:   envinfo.INGRESS,
+				},
+			},
+			returnedRoutes:  &routev1.RouteList{},
+			returnedIngress: fake.GetIngressListWithMultiple("wildfly"),
+			createdURLs:     []URL{},
+			deletedURLs:     []URL{},
 		},
 		{
 			name:          "2 (1 ingress,1 route) urls on env file and 2 on openshift cluster (1 ingress,1 route), but they are different",
@@ -1371,6 +1439,16 @@ func TestPush(t *testing.T) {
 				if !tt.args.isRouteSupported {
 					if len(fakeClientSet.RouteClientset.Actions()) > 0 {
 						t.Errorf("route is not supproted, total actions on the routeClient should be 0")
+					}
+				}
+
+				if len(tt.createdURLs) == 0 && len(tt.deletedURLs) == 0 {
+					if len(fakeClientSet.RouteClientset.Actions()) > 1 {
+						t.Errorf("when urls are in sync, total action for route client set should be less than 1")
+					}
+
+					if len(fakeClientSet.Kubernetes.Actions()) > 1 {
+						t.Errorf("when urls are in snyc, total action for kubernetes client set should be less than 1")
 					}
 				}
 			}

--- a/tests/integration/devfile/cmd_devfile_push_test.go
+++ b/tests/integration/devfile/cmd_devfile_push_test.go
@@ -345,7 +345,6 @@ var _ = Describe("odo devfile push command tests", func() {
 			}
 			Expect(volumesMatched).To(Equal(true))
 		})
-
 	})
 
 })

--- a/tests/integration/devfile/cmd_devfile_url_test.go
+++ b/tests/integration/devfile/cmd_devfile_url_test.go
@@ -168,12 +168,20 @@ var _ = Describe("odo devfile url command tests", func() {
 			helper.CmdShouldPass("odo", "url", "create", url1)
 
 			helper.CmdShouldPass("odo", "push", "--devfile", "devfile.yaml", "--namespace", namespace)
+			pushStdOut := helper.CmdShouldPass("odo", "push", "--devfile", "devfile.yaml", "--namespace", namespace)
+			Expect(pushStdOut).NotTo(ContainSubstring("successfully deleted"))
+			Expect(pushStdOut).NotTo(ContainSubstring("created"))
+			Expect(pushStdOut).To(ContainSubstring("URLs are synced with the cluster, no changes are required"))
 
 			output := helper.CmdShouldPass("oc", "get", "routes", "--namespace", namespace)
 			Expect(output).Should(ContainSubstring(url1))
 
 			helper.CmdShouldPass("odo", "url", "delete", url1, "-f")
 			helper.CmdShouldPass("odo", "push", "--devfile", "devfile.yaml", "--namespace", namespace)
+			pushStdOut = helper.CmdShouldPass("odo", "push", "--devfile", "devfile.yaml", "--namespace", namespace)
+			Expect(pushStdOut).NotTo(ContainSubstring("successfully deleted"))
+			Expect(pushStdOut).NotTo(ContainSubstring("created"))
+			Expect(pushStdOut).To(ContainSubstring("URLs are synced with the cluster, no changes are required"))
 
 			output = helper.CmdShouldPass("oc", "get", "routes", "--namespace", namespace)
 			Expect(output).ShouldNot(ContainSubstring(url1))
@@ -193,6 +201,35 @@ var _ = Describe("odo devfile url command tests", func() {
 
 			output := helper.CmdShouldPass("oc", "get", "routes", "--namespace", namespace)
 			Expect(output).Should(ContainSubstring(url1))
+		})
+
+		It("should be able to push again twice after creating and deleting a url", func() {
+			var stdOut string
+			url1 := helper.RandString(5)
+			host := helper.RandString(5) + ".com"
+
+			helper.CmdShouldPass("git", "clone", "https://github.com/che-samples/web-nodejs-sample.git", projectDirPath)
+			helper.Chdir(projectDirPath)
+
+			helper.CmdShouldPass("odo", "create", "nodejs", "--project", namespace, componentName)
+
+			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs"), projectDirPath)
+
+			helper.CmdShouldPass("odo", "url", "create", url1, "--port", "3000", "--host", host, "--ingress")
+
+			helper.CmdShouldPass("odo", "push", "--devfile", "devfile.yaml", "--project", namespace)
+			stdOut = helper.CmdShouldPass("odo", "push", "--devfile", "devfile.yaml", "--project", namespace)
+			Expect(stdOut).NotTo(ContainSubstring("successfully deleted"))
+			Expect(stdOut).NotTo(ContainSubstring("created"))
+			Expect(stdOut).To(ContainSubstring("URLs are synced with the cluster, no changes are required"))
+
+			helper.CmdShouldPass("odo", "url", "delete", url1, "-f")
+
+			helper.CmdShouldPass("odo", "push", "--devfile", "devfile.yaml", "--project", namespace)
+			stdOut = helper.CmdShouldPass("odo", "push", "--devfile", "devfile.yaml", "--project", namespace)
+			Expect(stdOut).NotTo(ContainSubstring("successfully deleted"))
+			Expect(stdOut).NotTo(ContainSubstring("created"))
+			Expect(stdOut).To(ContainSubstring("URLs are synced with the cluster, no changes are required"))
 		})
 	})
 

